### PR TITLE
Updated erc20 contract to support DSL

### PR DESCRIPTION
**Pull request by darthsiroftardis**
Updated the ERC20 contract to support DSL and contract headers

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 [workspace]
 
 members = [
-    "contract",
-    "logic",
-    "tests"
+	"contract",
```
Comment (zie1ony) [inline Cargo.toml:4]: Use 4 whitespaces.

File: contract/Cargo.toml [hunk 2]
```diff
@@ -1,19 +1,25 @@
 [package]
 name = "contract"
-version = "0.1.1"
-authors = ["Maciej Zielinski <maciej@casperlabs.io>"]
+version = "0.1.0"
+authors = ["zie1ony <maciej@casperlabs.io>"]
```
Comment (zie1ony) [inline contract/Cargo.toml:4]: I prefer to have `Maciej Zielinski` here.

File: contract/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    contracts::{EntryPoint, EntryPointAccess, EntryPointType, EntryPoints},
+    runtime_args, CLType, CLTyped, CLValue, Group, Parameter, RuntimeArgs, URef, U256,
+};
+
+#[casperlabs_contract]
+mod ERC20 {
+
+    #[casperlabs_constructor]
+    fn constructor(tokenName: String, tokenSymbol2: String, tokenTotalSupply: U256) {
+        let _name: String = tokenName;
```
Comment (zie1ony) [inline contract/src/main.rs:31]: This construction:
```
        let _name: String = tokenName;
        set_key("_name", _name);
```
Can be simplified:
```
set_key("_name", tokenName);
```

File: contract/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            runtime::get_caller(),
+            (get_key::<U256>(&new_key(
+                &new_key("_allowances", owner),
+                runtime::get_caller(),
+            )) - amount),
+        );
+    }
+
+    fn _transfer(sender: AccountHash, recipient: AccountHash, amount: U256) {
+        let temp4: U256 = (get_key::<U256>(&new_key("_balances", sender)) - amount);
```
Comment (zie1ony) [inline contract/src/main.rs:100]: Rename `temps`, or get ride of them in general.

File: tests/src/erc20.rs [hunk 5]
```diff
@@ -1,152 +1,126 @@
...
-    fn contract_hash(context: &TestContext, name: &str) -> Hash {
-        let contract_ref: Key = context
-            .query(account::ALI, &[name])
-            .unwrap_or_else(|_| panic!("{} contract not found.", name))
-            .into_t()
-            .unwrap_or_else(|_| panic!("{} is not a type Contract.", name));
-        contract_ref
-            .into_hash()
-            .unwrap_or_else(|| panic!("{} is not a type Hash.", name))
+    pub fn allowance(&self, owner: AccountHash, spender: AccountHash) -> U256 {
```
Comment (zie1ony) [inline tests/src/erc20.rs:100]: Reuse this function for allowances keys in contracts.
Do the same for balances.

File: contract/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    bytesrepr::{FromBytes, ToBytes},
+    contracts::{EntryPoint, EntryPointAccess, EntryPointType, EntryPoints},
+    runtime_args, CLType, CLTyped, CLValue, Group, Parameter, RuntimeArgs, URef, U256,
+};
+
+#[casperlabs_contract]
+mod ERC20 {
+
+    #[casperlabs_constructor]
+    fn constructor(tokenName: String, tokenSymbol2: String, tokenTotalSupply: U256) {
```
Comment (zie1ony) [inline contract/src/main.rs:30]: `tokenSymbol2`, why `2`?

File: contract/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        set_key("_decimals", _decimals);
+        let temp8: U256 = tokenTotalSupply;
+        set_key(&new_key("_balances", runtime::get_caller()), temp8);
+        let _totalSupply: U256 = tokenTotalSupply;
+        set_key("_totalSupply", _totalSupply);
+    }
+
+    #[casperlabs_method]
+    fn name() {
+        ret(get_key::<String>("_name"));
```
Comment (zie1ony) [inline contract/src/main.rs:45]: @darthsiroftardis haven't you implemented `returns` in DLS?

File: contract/src/main.rs (hunk 6 above)
Comment (darthsiroftardis) [reply contract/src/main.rs:30]: Renamed to tokenSymbol

File: contract/src/main.rs (hunk 3 above)
Comment (darthsiroftardis) [reply contract/src/main.rs:31]: Done

File: contract/Cargo.toml (hunk 2 above)
Comment (darthsiroftardis) [reply contract/Cargo.toml:4]: Renamed, sorry for that

File: contract/src/main.rs (hunk 7 above)
Comment (darthsiroftardis) [reply contract/src/main.rs:45]: I did, it pointed to a previous version

File: contract/src/main.rs (hunk 3 above)
Comment (zie1ony) [reply contract/src/main.rs]: Not everywhere.

File: tests/src/erc20.rs (hunk 5 above)
Comment (zie1ony) [reply tests/src/erc20.rs:100]: That's not done.

Comment (darthsiroftardis) [reply tests/src/erc20.rs:100]: The one in tests uses `self.query_contract` which I don't understand how we would reuse it?

File: Cargo.toml (hunk 1 above)
Comment (darthsiroftardis) [reply Cargo.toml]: Used 4 whitespaces

File: tests/src/erc20.rs (hunk 5 above)
Comment (darthsiroftardis) [reply tests/src/erc20.rs:99]: Changed the keys in main

File: contract/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+#[casperlabs_contract]
+mod ERC20 {
+
+    #[casperlabs_constructor]
+    fn constructor(tokenName: String, tokenSymbol: String, tokenTotalSupply: U256) {
+        set_key("_name", tokenName);
+        set_key("_symbol", tokenSymbol);
+        set_key("_decimals", 18u8);
+        let balances_key = format!("_balances_{}", runtime::get_caller());
```
Comment (zie1ony) [inline contract/src/main.rs:34]: Rename to `balance` It's a balance of a single account.

File: contract/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        set_key("_totalSupply", tokenTotalSupply);
+    }
+
+    #[casperlabs_method]
+    fn name() {
+        ret(get_key::<String>("_name"));
+    }
+
+    #[casperlabs_method]
+    fn symbol() {
```
Comment (zie1ony) [inline contract/src/main.rs:45]: Remove the `ret` and return normaly. I can also remove `<String>` if you have return type defined.

File: contract/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        ret(get_key::<String>("_symbol"));
+    }
+
+    #[casperlabs_method]
+    fn decimals() {
+        ret(get_key::<u8>("_decimals"));
```
Comment (zie1ony) [inline contract/src/main.rs:51]: Remove the `ret` and return normaly.

File: contract/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    #[casperlabs_method]
+    fn totalSupply() {
+        ret(get_key::<U256>("_totalSupply"));
```
Comment (zie1ony) [inline contract/src/main.rs:56]: Remove the `ret` and return normaly.

File: contract/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    #[casperlabs_method]
+    fn transfer(recipient: AccountHash, amount: U256) {
+        _transfer(runtime::get_caller(), recipient, amount);
+    }
+
+    #[casperlabs_method]
+    fn balance_of(account: AccountHash) {
```
Comment (zie1ony) [inline contract/src/main.rs:65]: This should look like:
```
get_key(balance_key(&account))
```

File: contract/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        let key = format!("_balances_{}", account);
+        ret(get_key::<U256>(&key));
+    }
+
+    #[casperlabs_method]
+    fn allowance(owner: AccountHash, spender: AccountHash) {
```
Comment (zie1ony) [inline contract/src/main.rs:71]: This should look like
```
get_key(allowance_key(&owner, &spender))
```

File: contract/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    #[casperlabs_method]
+    fn transferFrom(owner: AccountHash, recipient: AccountHash, amount: U256) {
+        let key = format!("_allowances_{}_{}", owner, runtime::get_caller());
+        _transfer(owner, recipient, amount);
+        _approve(
+            owner,
+            runtime::get_caller(),
+            (get_key::<U256>(&key) - amount),
```
Comment (zie1ony) [inline contract/src/main.rs:88]: same here

File: contract/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        );
+    }
+
+    fn _transfer(sender: AccountHash, recipient: AccountHash, amount: U256) {
+        let sender_key = format!("_balances_{}", sender);
```
Comment (zie1ony) [inline contract/src/main.rs:93]: same here

File: contract/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        let recipient_key = format!("_balances_{}", recipient);
```
Comment (zie1ony) [inline contract/src/main.rs:94]: same here

File: contract/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        let new_sender_balance: U256 = (get_key::<U256>(&sender_key) - amount);
+        set_key(&sender_key, new_sender_balance);
+        let new_recipient_balance: U256 = (get_key::<U256>(&recipient_key) + amount);
+        set_key(&recipient_key, new_recipient_balance);
+    }
+
+    fn _approve(owner: AccountHash, spender: AccountHash, amount: U256) {
+        let key = format!("_allowances_{}_{}", owner, spender);
```
Comment (zie1ony) [inline contract/src/main.rs:102]: same here

**Pull request merged**